# Feature Selection-Filter Methods

Basic methods : constant features, quasi constant feature , duplicate feature


Correlation method


## Constant features -Santander Customer Satisfaction Kaggle
Constant features are those that show the same value, just one value, for all the observations of the dataset. This is, the same value for all the rows of the dataset. These features provide no information that allows a machine learning model to discriminate or predict a target.

In [60]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.feature_selection import VarianceThreshold

In [61]:
data = pd.read_csv('D:/ShakthiSai -Keerthana Learnings/AI and Machine Learning/Feature Engineering/santander-customer-satisfaction/Santander Customer Satisfaction-train.csv', nrows=50000)
data.shape

(50000, 371)

In [62]:
data.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   
3                      0.0                      0.0  ...   
4                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [63]:
data.isnull().sum()

ID                         0
var3                       0
var15                      0
imp_ent_var16_ult1         0
imp_op_var39_comer_ult1    0
                          ..
saldo_medio_var44_hace3    0
saldo_medio_var44_ult1     0
saldo_medio_var44_ult3     0
var38                      0
TARGET                     0
Length: 371, dtype: int64

In [64]:
X=data.iloc[:,:-1]
y=data.iloc[:,-1]

In [65]:
X.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   
3                      0.0                      0.0  ...   
4                      0.0                      0.0  ...   

   saldo_medio_var29_ult3  saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
0                     0.0                     0.0                      0.0   
1                     0.0                     0.0                      0.0   
2                     0.0                     0.0                      0.0   
3                     0.0                     0.0                      0.0   
4                     0.0                     0.0                      0.0   

   saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
0                      0.0                     0.0                     0.0   
1                      0.0                     0.0                     0.0   
2                      0.0                     0.0                     0.0   
3                      0.0                     0.0                     0.0   
4                      0.0                     0.0                     0.0   

           var38  
0   39205.170000  
1   49278.030000  
2   67333.770000  
3   64007.970000  
4  117310.979016  

[5 rows x 370 columns]

In [66]:
y

0        0
1        0
2        0
3        0
4        0
        ..
49995    0
49996    0
49997    1
49998    0
49999    0
Name: TARGET, Length: 50000, dtype: int64

In [67]:
X_train, X_test, y_train, y_test = train_test_split(
    X,y,
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 370), (15000, 370))

## Method 1 : Using variance threshold from sklearn

It removes all features which variance doesn’t meet some threshold. By default, it removes all zero-variance features, i.e., features that have the same value in all samples.

In [68]:
cons = VarianceThreshold(threshold=0)
cons.fit(X_train)

VarianceThreshold(threshold=0)

In [69]:
cons.get_support()  ## get_support is a boolean vector that indicates which features are retained

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False, False,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False, False, False, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [70]:
sum(cons.get_support())

312

In [71]:
# constant features

[x for x in X_train.columns if x not in X_train.columns[cons.get_support()]]

['ind_var2_0',
 'ind_var2',
 'ind_var13_medio_0',
 'ind_var13_medio',
 'ind_var27_0',
 'ind_var28_0',
 'ind_var28',
 'ind_var27',
 'ind_var34_0',
 'ind_var34',
 'ind_var41',
 'ind_var46_0',
 'ind_var46',
 'num_var13_medio_0',
 'num_var13_medio',
 'num_var27_0',
 'num_var28_0',
 'num_var28',
 'num_var27',
 'num_var34_0',
 'num_var34',
 'num_var41',
 'num_var46_0',
 'num_var46',
 'saldo_var13_medio',
 'saldo_var28',
 'saldo_var27',
 'saldo_var34',
 'saldo_var41',
 'saldo_var46',
 'delta_imp_amort_var34_1y3',
 'delta_imp_reemb_var33_1y3',
 'delta_num_reemb_var33_1y3',
 'imp_amort_var18_hace3',
 'imp_amort_var34_hace3',
 'imp_amort_var34_ult1',
 'imp_reemb_var13_hace3',
 'imp_reemb_var17_hace3',
 'imp_reemb_var33_hace3',
 'imp_reemb_var33_ult1',
 'imp_trasp_var17_out_hace3',
 'imp_trasp_var33_out_hace3',
 'imp_venta_var44_hace3',
 'num_var2_0_ult1',
 'num_var2_ult1',
 'num_meses_var13_medio_ult3',
 'num_reemb_var13_hace3',
 'num_reemb_var17_hace3',
 'num_reemb_var33_hace3',
 'num_reemb_var

In [72]:
len([x for x in X_train.columns if x not in X_train.columns[cons.get_support()]])

58

In [73]:
## 58 variables are constant features..i.e feature with same value for all rows

In [74]:
data['saldo_var2_ult1']

0        0
1        0
2        0
3        0
4        0
        ..
49995    0
49996    0
49997    0
49998    0
49999    0
Name: saldo_var2_ult1, Length: 50000, dtype: int64

In [75]:
# transform function to reduce the training and testing sets

In [76]:
X_train = cons.transform(X_train)
X_test = cons.transform(X_test)

X_train.shape, X_test.shape

((35000, 312), (15000, 312))

## Method 2 : General Coding

In [77]:
data1 = pd.read_csv('D:/ShakthiSai -Keerthana Learnings/AI and Machine Learning/Feature Engineering/santander-customer-satisfaction/Santander Customer Satisfaction-train.csv', nrows=50000)
data1.shape

(50000, 371)

In [78]:
X=data.iloc[:,:-1]
y=data.iloc[:,-1]

In [79]:
X_train, X_test, y_train, y_test = train_test_split(
    X,y,
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 370), (15000, 370))

In [80]:
# to find constant features, use its std=0
constant_features = [
    feat for feat in X_train.columns if X_train[feat].std() == 0
]

len(constant_features)

58

In [81]:
# we can then drop these columns from the train and test sets
X_train.drop(labels=constant_features, axis=1, inplace=True)
X_test.drop(labels=constant_features, axis=1, inplace=True)

X_train.shape, X_test.shape

C:\Users\Vikee\anaconda3\lib\site-packages\pandas\core\frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


((35000, 312), (15000, 312))

## Quasi-Constant features -Santander Customer Satisfaction Kaggle


Quasi-constant features are those that show the same value for the great majority of the observations of the dataset. 
Quasi-constant features, as the name suggests, are the features that are almost constant. In other words, these features have the same values for a very large subset of the outputs. Such features are not very useful for making predictions.

## Method 1 : Using variance threshold from sklearn

In [82]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.feature_selection import VarianceThreshold

In [83]:
df = pd.read_csv('D:/ShakthiSai -Keerthana Learnings/AI and Machine Learning/Feature Engineering/santander-customer-satisfaction/Santander Customer Satisfaction-train.csv', nrows=50000)
df.shape

(50000, 371)

In [85]:
df.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   
3                      0.0                      0.0  ...   
4                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [84]:
df.isnull().sum()

ID                         0
var3                       0
var15                      0
imp_ent_var16_ult1         0
imp_op_var39_comer_ult1    0
                          ..
saldo_medio_var44_hace3    0
saldo_medio_var44_ult1     0
saldo_medio_var44_ult3     0
var38                      0
TARGET                     0
Length: 371, dtype: int64

In [86]:
X=data.iloc[:,:-1]
y=data.iloc[:,-1]

In [87]:
X_train, X_test, y_train, y_test = train_test_split(
    X,y,
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 370), (15000, 370))

In [88]:
qua = VarianceThreshold(
    threshold=0.01)  # 0.1 indicates 99% of observations approximately

qua.fit(X_train)

VarianceThreshold(threshold=0.01)

In [89]:
sum(qua.get_support())

262

In [90]:
[x for x in X_train.columns if x not in X_train.columns[qua.get_support()]]

['ind_var1',
 'ind_var2_0',
 'ind_var2',
 'ind_var6_0',
 'ind_var6',
 'ind_var13_medio_0',
 'ind_var13_medio',
 'ind_var14',
 'ind_var17_0',
 'ind_var17',
 'ind_var18_0',
 'ind_var18',
 'ind_var19',
 'ind_var20_0',
 'ind_var20',
 'ind_var27_0',
 'ind_var28_0',
 'ind_var28',
 'ind_var27',
 'ind_var29_0',
 'ind_var29',
 'ind_var30_0',
 'ind_var31_0',
 'ind_var31',
 'ind_var32_cte',
 'ind_var32_0',
 'ind_var32',
 'ind_var33_0',
 'ind_var33',
 'ind_var34_0',
 'ind_var34',
 'ind_var40',
 'ind_var41',
 'ind_var39',
 'ind_var44_0',
 'ind_var44',
 'ind_var46_0',
 'ind_var46',
 'num_var6_0',
 'num_var6',
 'num_var13_medio_0',
 'num_var13_medio',
 'num_var18_0',
 'num_var18',
 'num_op_var40_hace3',
 'num_var27_0',
 'num_var28_0',
 'num_var28',
 'num_var27',
 'num_var29_0',
 'num_var29',
 'num_var33',
 'num_var34_0',
 'num_var34',
 'num_var41',
 'num_var46_0',
 'num_var46',
 'saldo_var13_medio',
 'saldo_var28',
 'saldo_var27',
 'saldo_var34',
 'saldo_var41',
 'saldo_var46',
 'delta_imp_amort_var3

In [91]:
X_train = qua.transform(X_train)
X_test = qua.transform(X_test)

X_train.shape, X_test.shape

((35000, 262), (15000, 262))

## Method 2 : General Coding

In [92]:
data2 = pd.read_csv('D:/ShakthiSai -Keerthana Learnings/AI and Machine Learning/Feature Engineering/santander-customer-satisfaction/Santander Customer Satisfaction-train.csv', nrows=50000)
data2.shape

(50000, 371)

In [93]:
X=data.iloc[:,:-1]
y=data.iloc[:,-1]

In [94]:
X_train, X_test, y_train, y_test = train_test_split(
    X,y,
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 370), (15000, 370))

In [95]:
# remove constant features
# using the code from the previous lecture
constant_features = [
    feat for feat in X_train.columns if X_train[feat].std() == 0
]

X_train.drop(labels=constant_features, axis=1, inplace=True)
X_test.drop(labels=constant_features, axis=1, inplace=True)

X_train.shape, X_test.shape

C:\Users\Vikee\anaconda3\lib\site-packages\pandas\core\frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


((35000, 312), (15000, 312))

In [96]:
# percentage of observations showing each of the different values
(X_train[feature].value_counts() / np.float(
        len(X_train)))

NameError: name 'feature' is not defined

In [97]:
# to find constant features, use its std=0
quasi_constant_feat = []
for feature in X_train.columns:

    # find the predominant value
    predominant = (X_train[feature].value_counts() / np.float(
        len(X_train))).sort_values(ascending=False).values[0]

    # evaluate predominant feature
    if predominant > 0.998:
        quasi_constant_feat.append(feature)

len(quasi_constant_feat)

119

In [98]:
# select the first one from the list
quasi_constant_feat[0]

'imp_op_var40_efect_ult1'

In [99]:
X_train['imp_op_var40_efect_ult1'].value_counts() / np.float(len(X_train))

0.00       0.999400
900.00     0.000086
60.00      0.000057
1800.00    0.000057
600.00     0.000057
930.00     0.000029
420.00     0.000029
74.28      0.000029
270.00     0.000029
1200.00    0.000029
6600.00    0.000029
870.00     0.000029
750.00     0.000029
300.00     0.000029
120.00     0.000029
210.00     0.000029
150.00     0.000029
Name: imp_op_var40_efect_ult1, dtype: float64

The feature shows 0 for more than 99.9% of the observations.

## Duplicated features -Santander Customer Satisfaction Kaggle

datasets contain one or more features that show the same values across all the observations. This means that both features are in essence identical. In addition, it is not unusual to introduce duplicated features after performing one hot encoding of categorical variables, particularly when using several highly cardinal variables.

Identifying and removing duplicated, and therefore redundant features, is an easy first step towards feature selection and more easily interpretable machine learning models.

There is no function in python and pandas to find duplicated columns.
we have to write our own code


## Method for small datasets

In [21]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

dataset = pd.read_csv('D:/ShakthiSai -Keerthana Learnings/AI and Machine Learning/Feature Engineering/santander-customer-satisfaction/Santander Customer Satisfaction-train.csv', nrows=15000)
dataset.shape

(15000, 371)

In [22]:
dataset.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                        0                        0  ...   
1                        0                        0  ...   
2                        0                        0  ...   
3                        0                        0  ...   
4                        0                        0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [23]:
dataset.isnull().sum()

ID                         0
var3                       0
var15                      0
imp_ent_var16_ult1         0
imp_op_var39_comer_ult1    0
                          ..
saldo_medio_var44_hace3    0
saldo_medio_var44_ult1     0
saldo_medio_var44_ult3     0
var38                      0
TARGET                     0
Length: 371, dtype: int64

In [100]:
X=dataset.iloc[:,:-1]
y=dataset.iloc[:,-1]

In [101]:
X_train, X_test, y_train, y_test = train_test_split(
    X,y,
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((10500, 370), (4500, 370))

Pandas has the function 'duplicated' that evaluates if the dataframe contains duplicated rows. We can use this function to check for duplicated columns if we transpose the dataframe first. By transposing the dataframe, we obtain a new dataframe where the columns are now rows, and with the 'duplicated' method we can go ahead an identify those that are duplicated.

Once we identify them, we can remove the duplicated rows. 

Using pandas transpose is computationally expensive, so the computer may run out of memory. That is why we can only use this code block on small datasets. How small will depend of your computer specifications.

In [102]:
dataset_t = X_train.T
dataset_t.head()

10439    9236    818      11504    11722    5276   \
ID                       20941.0  18583.0  1623.0  23060.0  23512.0  10564.0   
var3                         2.0      2.0     2.0      2.0      2.0      2.0   
var15                       23.0     39.0    22.0     23.0     37.0     23.0   
imp_ent_var16_ult1           0.0      0.0   150.0      0.0      0.0      0.0   
imp_op_var39_comer_ult1      0.0      0.0     0.0      0.0      0.0      0.0   

                           6863     13463    10228    11462  ...   4373   \
ID                       13779.0  26969.0  20502.0  22981.0  ...  8783.0   
var3                         2.0      2.0      2.0      2.0  ...     2.0   
var15                       27.0     43.0     23.0     27.0  ...    23.0   
imp_ent_var16_ult1           0.0      0.0      0.0      0.0  ...     0.0   
imp_op_var39_comer_ult1      0.0      0.0      0.0      0.0  ...     0.0   

                           7891     9225     14019   4859     13123   3264   \
ID                       15901.0  18564.0  28142.0  9723.0  26306.0  6557.0   
var3                         2.0      2.0      2.0     2.0      2.0     2.0   
var15                       24.0     33.0     45.0    24.0     37.0    24.0   
imp_ent_var16_ult1           0.0      0.0      0.0     0.0      0.0     0.0   
imp_op_var39_comer_ult1      0.0      0.0      0.0     0.0      0.0     0.0   

                           9845     10799   2732   
ID                       19796.0  21653.0  5441.0  
var3                         2.0      2.0     2.0  
var15                       38.0     28.0    23.0  
imp_ent_var16_ult1           0.0      0.0     0.0  
imp_op_var39_comer_ult1      0.0      0.0     0.0  

[5 rows x 10500 columns]

In [103]:
dataset_t.duplicated().sum()

105

We can see that 105 columns / variables are duplicated. This means that 105 variables are identical to at least another variable within a dataset.

In [104]:
# visualise the duplicated rows (the columns of the original dataframe)
dataset_t[dataset_t.duplicated()]

10439  9236   818    11504  11722  5276   \
ind_var2                         0.0    0.0    0.0    0.0    0.0    0.0   
ind_var13_medio_0                0.0    0.0    0.0    0.0    0.0    0.0   
ind_var13_medio                  0.0    0.0    0.0    0.0    0.0    0.0   
ind_var18_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var18                        0.0    0.0    0.0    0.0    0.0    0.0   
...                              ...    ...    ...    ...    ...    ...   
saldo_medio_var13_medio_hace2    0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var13_medio_hace3    0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var13_medio_ult1     0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var13_medio_ult3     0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var29_hace3          0.0    0.0    0.0    0.0    0.0    0.0   

                               6863   13463  10228  11462  ...  4373   7891   \
ind_var2                         0.0    0.0    0.0    0.0  ...    0.0    0.0   
ind_var13_medio_0                0.0    0.0    0.0    0.0  ...    0.0    0.0   
ind_var13_medio                  0.0    0.0    0.0    0.0  ...    0.0    0.0   
ind_var18_0                      0.0    0.0    0.0    0.0  ...    0.0    0.0   
ind_var18                        0.0    0.0    0.0    0.0  ...    0.0    0.0   
...                              ...    ...    ...    ...  ...    ...    ...   
saldo_medio_var13_medio_hace2    0.0    0.0    0.0    0.0  ...    0.0    0.0   
saldo_medio_var13_medio_hace3    0.0    0.0    0.0    0.0  ...    0.0    0.0   
saldo_medio_var13_medio_ult1     0.0    0.0    0.0    0.0  ...    0.0    0.0   
saldo_medio_var13_medio_ult3     0.0    0.0    0.0    0.0  ...    0.0    0.0   
saldo_medio_var29_hace3          0.0    0.0    0.0    0.0  ...    0.0    0.0   

                               9225   14019  4859   13123  3264   9845   \
ind_var2                         0.0    0.0    0.0    0.0    0.0    0.0   
ind_var13_medio_0                0.0    0.0    0.0    0.0    0.0    0.0   
ind_var13_medio                  0.0    0.0    0.0    0.0    0.0    0.0   
ind_var18_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var18                        0.0    0.0    0.0    0.0    0.0    0.0   
...                              ...    ...    ...    ...    ...    ...   
saldo_medio_var13_medio_hace2    0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var13_medio_hace3    0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var13_medio_ult1     0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var13_medio_ult3     0.0    0.0    0.0    0.0    0.0    0.0   
saldo_medio_var29_hace3          0.0    0.0    0.0    0.0    0.0    0.0   

                               10799  2732   
ind_var2                         0.0    0.0  
ind_var13_medio_0                0.0    0.0  
ind_var13_medio                  0.0    0.0  
ind_var18_0                      0.0    0.0  
ind_var18                        0.0    0.0  
...                              ...    ...  
saldo_medio_var13_medio_hace2    0.0    0.0  
saldo_medio_var13_medio_hace3    0.0    0.0  
saldo_medio_var13_medio_ult1     0.0    0.0  
saldo_medio_var13_medio_ult3     0.0    0.0  
saldo_medio_var29_hace3          0.0    0.0  

[105 rows x 10500 columns]

In [105]:
# we can capture the duplicated features, by capturing the
# index values of the transposed dataframe like this:
duplicated_features = dataset_t[dataset_t.duplicated()].index.values
duplicated_features

array(['ind_var2', 'ind_var13_medio_0', 'ind_var13_medio', 'ind_var18_0',
       'ind_var18', 'ind_var26', 'ind_var25', 'ind_var27_0',
       'ind_var28_0', 'ind_var28', 'ind_var27', 'ind_var29_0',
       'ind_var29', 'ind_var32', 'ind_var34_0', 'ind_var34', 'ind_var37',
       'ind_var40_0', 'ind_var40', 'ind_var41', 'ind_var39', 'ind_var44',
       'ind_var46_0', 'ind_var46', 'num_var13_medio_0', 'num_var13_medio',
       'num_var18_0', 'num_var18', 'num_var26', 'num_var25',
       'num_var27_0', 'num_var28_0', 'num_var28', 'num_var27',
       'num_var29_0', 'num_var29', 'num_var32', 'num_var34_0',
       'num_var34', 'num_var37', 'num_var40_0', 'num_var40', 'num_var41',
       'num_var39', 'num_var46_0', 'num_var46', 'saldo_var13_medio',
       'saldo_var18', 'saldo_var28', 'saldo_var27', 'saldo_var29',
       'saldo_var34', 'saldo_var40', 'saldo_var41', 'saldo_var46',
       'delta_imp_amort_var18_1y3', 'delta_imp_amort_var34_1y3',
       'delta_imp_reemb_var17_1y3', 'delta_imp_ree

In [106]:
# alternatively, we can remove the duplicated rows,
# transpose the dataframe back to the variables as columns
# keep first indicates that we keep the first of a set of
# duplicated variables

dataset_unique = dataset_t.drop_duplicates(keep='first').T
dataset_unique.shape

(10500, 265)

We can see immediately how removing duplicated features helps reduce the feature space. We passed from 370 to 265 non-duplicated features.

In [107]:
# to find those columns in the original dataframe that were removed:

duplicated_features = [col for col in dataset.columns if col not in dataset_unique.columns]
duplicated_features

['ind_var2',
 'ind_var13_medio_0',
 'ind_var13_medio',
 'ind_var18_0',
 'ind_var18',
 'ind_var26',
 'ind_var25',
 'ind_var27_0',
 'ind_var28_0',
 'ind_var28',
 'ind_var27',
 'ind_var29_0',
 'ind_var29',
 'ind_var32',
 'ind_var34_0',
 'ind_var34',
 'ind_var37',
 'ind_var40_0',
 'ind_var40',
 'ind_var41',
 'ind_var39',
 'ind_var44',
 'ind_var46_0',
 'ind_var46',
 'num_var13_medio_0',
 'num_var13_medio',
 'num_var18_0',
 'num_var18',
 'num_var26',
 'num_var25',
 'num_var27_0',
 'num_var28_0',
 'num_var28',
 'num_var27',
 'num_var29_0',
 'num_var29',
 'num_var32',
 'num_var34_0',
 'num_var34',
 'num_var37',
 'num_var40_0',
 'num_var40',
 'num_var41',
 'num_var39',
 'num_var46_0',
 'num_var46',
 'saldo_var13_medio',
 'saldo_var18',
 'saldo_var28',
 'saldo_var27',
 'saldo_var29',
 'saldo_var34',
 'saldo_var40',
 'saldo_var41',
 'saldo_var46',
 'delta_imp_amort_var18_1y3',
 'delta_imp_amort_var34_1y3',
 'delta_imp_reemb_var17_1y3',
 'delta_imp_reemb_var33_1y3',
 'delta_imp_trasp_var17_out_1y3

## Method for large or big  datasets

Transposing a dataframe is memory costly if the dataframe is big. Therefore, we can use the alternative loop to find duplicated columns in bigger datasets.

In [108]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

dataset1 = pd.read_csv('D:/ShakthiSai -Keerthana Learnings/AI and Machine Learning/Feature Engineering/santander-customer-satisfaction/Santander Customer Satisfaction-train.csv')
dataset1.shape

(76020, 371)

In [112]:
X=dataset.iloc[:,:-1]
y=dataset.iloc[:,-1]

In [113]:
X_train, X_test, y_train, y_test = train_test_split(
    X,y,
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((10500, 370), (4500, 370))

In [114]:
# check for duplicated features in the training set
duplicated_feat = []
for i in range(0, len(X_train.columns)):

    col_1 = X_train.columns[i]

    for col_2 in X_train.columns[i + 1:]:
        if X_train[col_1].equals(X_train[col_2]):
            duplicated_feat.append(col_2)

In [115]:
# check how many features are duplicated
print(len(set(duplicated_feat)))

104


In [116]:
# let's print the list of duplicated features
set(duplicated_feat)

{'delta_imp_amort_var18_1y3',
 'delta_imp_amort_var34_1y3',
 'delta_imp_reemb_var17_1y3',
 'delta_imp_reemb_var33_1y3',
 'delta_imp_trasp_var17_out_1y3',
 'delta_num_aport_var17_1y3',
 'delta_num_aport_var33_1y3',
 'delta_num_reemb_var13_1y3',
 'delta_num_reemb_var17_1y3',
 'delta_num_reemb_var33_1y3',
 'delta_num_trasp_var17_in_1y3',
 'delta_num_trasp_var17_out_1y3',
 'delta_num_trasp_var33_in_1y3',
 'delta_num_trasp_var33_out_1y3',
 'imp_amort_var18_hace3',
 'imp_amort_var18_ult1',
 'imp_amort_var34_hace3',
 'imp_amort_var34_ult1',
 'imp_aport_var33_ult1',
 'imp_reemb_var13_hace3',
 'imp_reemb_var17_hace3',
 'imp_reemb_var33_hace3',
 'imp_reemb_var33_ult1',
 'imp_trasp_var17_out_hace3',
 'imp_trasp_var17_out_ult1',
 'imp_trasp_var33_in_hace3',
 'imp_trasp_var33_out_hace3',
 'imp_var7_emit_ult1',
 'ind_var13_medio',
 'ind_var13_medio_0',
 'ind_var18',
 'ind_var18_0',
 'ind_var2',
 'ind_var25',
 'ind_var26',
 'ind_var27',
 'ind_var27_0',
 'ind_var28',
 'ind_var28_0',
 'ind_var29',
 'in

In [117]:
# we can go ahead and try to identify which set of features
# are identical

duplicated_feat = []
for i in range(0, len(X_train.columns)):

    col_1 = X_train.columns[i]

    for col_2 in X_train.columns[i + 1:]:

        # if the features are duplicated
        if X_train[col_1].equals(X_train[col_2]):

            #print them
            print(col_1)
            print(col_2)
            print()

            # and then append the duplicated one to a
            # list
            duplicated_feat.append(col_2)

ind_var1_0
ind_var40_0

ind_var1
ind_var40

ind_var1
ind_var39

ind_var2_0
ind_var2

ind_var2_0
ind_var13_medio_0

ind_var2_0
ind_var13_medio

ind_var2_0
ind_var18_0

ind_var2_0
ind_var18

ind_var2_0
ind_var27_0

ind_var2_0
ind_var28_0

ind_var2_0
ind_var28

ind_var2_0
ind_var27

ind_var2_0
ind_var34_0

ind_var2_0
ind_var34

ind_var2_0
ind_var41

ind_var2_0
ind_var46_0

ind_var2_0
ind_var46

ind_var2_0
num_var13_medio_0

ind_var2_0
num_var13_medio

ind_var2_0
num_var18_0

ind_var2_0
num_var18

ind_var2_0
num_var27_0

ind_var2_0
num_var28_0

ind_var2_0
num_var28

ind_var2_0
num_var27

ind_var2_0
num_var34_0

ind_var2_0
num_var34

ind_var2_0
num_var41

ind_var2_0
num_var46_0

ind_var2_0
num_var46

ind_var2_0
saldo_var13_medio

ind_var2_0
saldo_var18

ind_var2_0
saldo_var28

ind_var2_0
saldo_var27

ind_var2_0
saldo_var34

ind_var2_0
saldo_var41

ind_var2_0
saldo_var46

ind_var2_0
delta_imp_amort_var18_1y3

ind_var2_0
delta_imp_amort_var34_1y3

ind_var2_0
delta_imp_reemb_var17_1y3

ind_var

ind_var13_medio
num_trasp_var33_in_hace3

ind_var13_medio
num_trasp_var33_out_hace3

ind_var13_medio
saldo_var2_ult1

ind_var13_medio
saldo_medio_var13_medio_hace2

ind_var13_medio
saldo_medio_var13_medio_hace3

ind_var13_medio
saldo_medio_var13_medio_ult1

ind_var13_medio
saldo_medio_var13_medio_ult3

ind_var13_medio
saldo_medio_var29_hace3

ind_var18_0
ind_var18

ind_var18_0
ind_var27_0

ind_var18_0
ind_var28_0

ind_var18_0
ind_var28

ind_var18_0
ind_var27

ind_var18_0
ind_var34_0

ind_var18_0
ind_var34

ind_var18_0
ind_var41

ind_var18_0
ind_var46_0

ind_var18_0
ind_var46

ind_var18_0
num_var13_medio_0

ind_var18_0
num_var13_medio

ind_var18_0
num_var18_0

ind_var18_0
num_var18

ind_var18_0
num_var27_0

ind_var18_0
num_var28_0

ind_var18_0
num_var28

ind_var18_0
num_var27

ind_var18_0
num_var34_0

ind_var18_0
num_var34

ind_var18_0
num_var41

ind_var18_0
num_var46_0

ind_var18_0
num_var46

ind_var18_0
saldo_var13_medio

ind_var18_0
saldo_var18

ind_var18_0
saldo_var28

ind_var18_0
s

ind_var28_0
num_meses_var13_medio_ult3

ind_var28_0
num_reemb_var13_hace3

ind_var28_0
num_reemb_var17_hace3

ind_var28_0
num_reemb_var17_ult1

ind_var28_0
num_reemb_var33_hace3

ind_var28_0
num_reemb_var33_ult1

ind_var28_0
num_trasp_var17_out_hace3

ind_var28_0
num_trasp_var17_out_ult1

ind_var28_0
num_trasp_var33_in_hace3

ind_var28_0
num_trasp_var33_out_hace3

ind_var28_0
saldo_var2_ult1

ind_var28_0
saldo_medio_var13_medio_hace2

ind_var28_0
saldo_medio_var13_medio_hace3

ind_var28_0
saldo_medio_var13_medio_ult1

ind_var28_0
saldo_medio_var13_medio_ult3

ind_var28_0
saldo_medio_var29_hace3

ind_var28
ind_var27

ind_var28
ind_var34_0

ind_var28
ind_var34

ind_var28
ind_var41

ind_var28
ind_var46_0

ind_var28
ind_var46

ind_var28
num_var13_medio_0

ind_var28
num_var13_medio

ind_var28
num_var18_0

ind_var28
num_var18

ind_var28
num_var27_0

ind_var28
num_var28_0

ind_var28
num_var28

ind_var28
num_var27

ind_var28
num_var34_0

ind_var28
num_var34

ind_var28
num_var41

ind_var28
num_

ind_var34
saldo_medio_var13_medio_hace2

ind_var34
saldo_medio_var13_medio_hace3

ind_var34
saldo_medio_var13_medio_ult1

ind_var34
saldo_medio_var13_medio_ult3

ind_var34
saldo_medio_var29_hace3

ind_var37_0
ind_var37

ind_var40
ind_var39

ind_var41
ind_var46_0

ind_var41
ind_var46

ind_var41
num_var13_medio_0

ind_var41
num_var13_medio

ind_var41
num_var18_0

ind_var41
num_var18

ind_var41
num_var27_0

ind_var41
num_var28_0

ind_var41
num_var28

ind_var41
num_var27

ind_var41
num_var34_0

ind_var41
num_var34

ind_var41
num_var41

ind_var41
num_var46_0

ind_var41
num_var46

ind_var41
saldo_var13_medio

ind_var41
saldo_var18

ind_var41
saldo_var28

ind_var41
saldo_var27

ind_var41
saldo_var34

ind_var41
saldo_var41

ind_var41
saldo_var46

ind_var41
delta_imp_amort_var18_1y3

ind_var41
delta_imp_amort_var34_1y3

ind_var41
delta_imp_reemb_var17_1y3

ind_var41
delta_imp_reemb_var33_1y3

ind_var41
delta_imp_trasp_var17_out_1y3

ind_var41
delta_num_reemb_var17_1y3

ind_var41
delta_num_reemb

num_var13_medio
num_var34

num_var13_medio
num_var41

num_var13_medio
num_var46_0

num_var13_medio
num_var46

num_var13_medio
saldo_var13_medio

num_var13_medio
saldo_var18

num_var13_medio
saldo_var28

num_var13_medio
saldo_var27

num_var13_medio
saldo_var34

num_var13_medio
saldo_var41

num_var13_medio
saldo_var46

num_var13_medio
delta_imp_amort_var18_1y3

num_var13_medio
delta_imp_amort_var34_1y3

num_var13_medio
delta_imp_reemb_var17_1y3

num_var13_medio
delta_imp_reemb_var33_1y3

num_var13_medio
delta_imp_trasp_var17_out_1y3

num_var13_medio
delta_num_reemb_var17_1y3

num_var13_medio
delta_num_reemb_var33_1y3

num_var13_medio
delta_num_trasp_var17_out_1y3

num_var13_medio
imp_amort_var18_hace3

num_var13_medio
imp_amort_var18_ult1

num_var13_medio
imp_amort_var34_hace3

num_var13_medio
imp_amort_var34_ult1

num_var13_medio
imp_aport_var33_ult1

num_var13_medio
imp_var7_emit_ult1

num_var13_medio
imp_reemb_var13_hace3

num_var13_medio
imp_reemb_var17_hace3

num_var13_medio
imp_ree

num_var28_0
num_reemb_var13_hace3

num_var28_0
num_reemb_var17_hace3

num_var28_0
num_reemb_var17_ult1

num_var28_0
num_reemb_var33_hace3

num_var28_0
num_reemb_var33_ult1

num_var28_0
num_trasp_var17_out_hace3

num_var28_0
num_trasp_var17_out_ult1

num_var28_0
num_trasp_var33_in_hace3

num_var28_0
num_trasp_var33_out_hace3

num_var28_0
saldo_var2_ult1

num_var28_0
saldo_medio_var13_medio_hace2

num_var28_0
saldo_medio_var13_medio_hace3

num_var28_0
saldo_medio_var13_medio_ult1

num_var28_0
saldo_medio_var13_medio_ult3

num_var28_0
saldo_medio_var29_hace3

num_var28
num_var27

num_var28
num_var34_0

num_var28
num_var34

num_var28
num_var41

num_var28
num_var46_0

num_var28
num_var46

num_var28
saldo_var13_medio

num_var28
saldo_var18

num_var28
saldo_var28

num_var28
saldo_var27

num_var28
saldo_var34

num_var28
saldo_var41

num_var28
saldo_var46

num_var28
delta_imp_amort_var18_1y3

num_var28
delta_imp_amort_var34_1y3

num_var28
delta_imp_reemb_var17_1y3

num_var28
delta_imp_reemb_var

num_var41
num_reemb_var33_hace3

num_var41
num_reemb_var33_ult1

num_var41
num_trasp_var17_out_hace3

num_var41
num_trasp_var17_out_ult1

num_var41
num_trasp_var33_in_hace3

num_var41
num_trasp_var33_out_hace3

num_var41
saldo_var2_ult1

num_var41
saldo_medio_var13_medio_hace2

num_var41
saldo_medio_var13_medio_hace3

num_var41
saldo_medio_var13_medio_ult1

num_var41
saldo_medio_var13_medio_ult3

num_var41
saldo_medio_var29_hace3

num_var46_0
num_var46

num_var46_0
saldo_var13_medio

num_var46_0
saldo_var18

num_var46_0
saldo_var28

num_var46_0
saldo_var27

num_var46_0
saldo_var34

num_var46_0
saldo_var41

num_var46_0
saldo_var46

num_var46_0
delta_imp_amort_var18_1y3

num_var46_0
delta_imp_amort_var34_1y3

num_var46_0
delta_imp_reemb_var17_1y3

num_var46_0
delta_imp_reemb_var33_1y3

num_var46_0
delta_imp_trasp_var17_out_1y3

num_var46_0
delta_num_reemb_var17_1y3

num_var46_0
delta_num_reemb_var33_1y3

num_var46_0
delta_num_trasp_var17_out_1y3

num_var46_0
imp_amort_var18_hace3

num_va

saldo_var28
num_meses_var13_medio_ult3

saldo_var28
num_reemb_var13_hace3

saldo_var28
num_reemb_var17_hace3

saldo_var28
num_reemb_var17_ult1

saldo_var28
num_reemb_var33_hace3

saldo_var28
num_reemb_var33_ult1

saldo_var28
num_trasp_var17_out_hace3

saldo_var28
num_trasp_var17_out_ult1

saldo_var28
num_trasp_var33_in_hace3

saldo_var28
num_trasp_var33_out_hace3

saldo_var28
saldo_var2_ult1

saldo_var28
saldo_medio_var13_medio_hace2

saldo_var28
saldo_medio_var13_medio_hace3

saldo_var28
saldo_medio_var13_medio_ult1

saldo_var28
saldo_medio_var13_medio_ult3

saldo_var28
saldo_medio_var29_hace3

saldo_var27
saldo_var34

saldo_var27
saldo_var41

saldo_var27
saldo_var46

saldo_var27
delta_imp_amort_var18_1y3

saldo_var27
delta_imp_amort_var34_1y3

saldo_var27
delta_imp_reemb_var17_1y3

saldo_var27
delta_imp_reemb_var33_1y3

saldo_var27
delta_imp_trasp_var17_out_1y3

saldo_var27
delta_num_reemb_var17_1y3

saldo_var27
delta_num_reemb_var33_1y3

saldo_var27
delta_num_trasp_var17_out_1y3

sa

delta_imp_amort_var18_1y3
saldo_medio_var13_medio_hace2

delta_imp_amort_var18_1y3
saldo_medio_var13_medio_hace3

delta_imp_amort_var18_1y3
saldo_medio_var13_medio_ult1

delta_imp_amort_var18_1y3
saldo_medio_var13_medio_ult3

delta_imp_amort_var18_1y3
saldo_medio_var29_hace3

delta_imp_amort_var34_1y3
delta_imp_reemb_var17_1y3

delta_imp_amort_var34_1y3
delta_imp_reemb_var33_1y3

delta_imp_amort_var34_1y3
delta_imp_trasp_var17_out_1y3

delta_imp_amort_var34_1y3
delta_num_reemb_var17_1y3

delta_imp_amort_var34_1y3
delta_num_reemb_var33_1y3

delta_imp_amort_var34_1y3
delta_num_trasp_var17_out_1y3

delta_imp_amort_var34_1y3
imp_amort_var18_hace3

delta_imp_amort_var34_1y3
imp_amort_var18_ult1

delta_imp_amort_var34_1y3
imp_amort_var34_hace3

delta_imp_amort_var34_1y3
imp_amort_var34_ult1

delta_imp_amort_var34_1y3
imp_aport_var33_ult1

delta_imp_amort_var34_1y3
imp_var7_emit_ult1

delta_imp_amort_var34_1y3
imp_reemb_var13_hace3

delta_imp_amort_var34_1y3
imp_reemb_var17_hace3

delta_imp_a

delta_imp_trasp_var33_in_1y3
delta_num_trasp_var33_in_1y3

delta_imp_trasp_var33_out_1y3
delta_num_trasp_var33_out_1y3

delta_num_reemb_var17_1y3
delta_num_reemb_var33_1y3

delta_num_reemb_var17_1y3
delta_num_trasp_var17_out_1y3

delta_num_reemb_var17_1y3
imp_amort_var18_hace3

delta_num_reemb_var17_1y3
imp_amort_var18_ult1

delta_num_reemb_var17_1y3
imp_amort_var34_hace3

delta_num_reemb_var17_1y3
imp_amort_var34_ult1

delta_num_reemb_var17_1y3
imp_aport_var33_ult1

delta_num_reemb_var17_1y3
imp_var7_emit_ult1

delta_num_reemb_var17_1y3
imp_reemb_var13_hace3

delta_num_reemb_var17_1y3
imp_reemb_var17_hace3

delta_num_reemb_var17_1y3
imp_reemb_var33_hace3

delta_num_reemb_var17_1y3
imp_reemb_var33_ult1

delta_num_reemb_var17_1y3
imp_trasp_var17_out_hace3

delta_num_reemb_var17_1y3
imp_trasp_var17_out_ult1

delta_num_reemb_var17_1y3
imp_trasp_var33_in_hace3

delta_num_reemb_var17_1y3
imp_trasp_var33_out_hace3

delta_num_reemb_var17_1y3
ind_var7_emit_ult1

delta_num_reemb_var17_1y3
num_v

imp_amort_var34_hace3
num_var7_emit_ult1

imp_amort_var34_hace3
num_meses_var13_medio_ult3

imp_amort_var34_hace3
num_reemb_var13_hace3

imp_amort_var34_hace3
num_reemb_var17_hace3

imp_amort_var34_hace3
num_reemb_var17_ult1

imp_amort_var34_hace3
num_reemb_var33_hace3

imp_amort_var34_hace3
num_reemb_var33_ult1

imp_amort_var34_hace3
num_trasp_var17_out_hace3

imp_amort_var34_hace3
num_trasp_var17_out_ult1

imp_amort_var34_hace3
num_trasp_var33_in_hace3

imp_amort_var34_hace3
num_trasp_var33_out_hace3

imp_amort_var34_hace3
saldo_var2_ult1

imp_amort_var34_hace3
saldo_medio_var13_medio_hace2

imp_amort_var34_hace3
saldo_medio_var13_medio_hace3

imp_amort_var34_hace3
saldo_medio_var13_medio_ult1

imp_amort_var34_hace3
saldo_medio_var13_medio_ult3

imp_amort_var34_hace3
saldo_medio_var29_hace3

imp_amort_var34_ult1
imp_aport_var33_ult1

imp_amort_var34_ult1
imp_var7_emit_ult1

imp_amort_var34_ult1
imp_reemb_var13_hace3

imp_amort_var34_ult1
imp_reemb_var17_hace3

imp_amort_var34_ult1
im

imp_reemb_var33_ult1
imp_trasp_var17_out_ult1

imp_reemb_var33_ult1
imp_trasp_var33_in_hace3

imp_reemb_var33_ult1
imp_trasp_var33_out_hace3

imp_reemb_var33_ult1
ind_var7_emit_ult1

imp_reemb_var33_ult1
num_var2_0_ult1

imp_reemb_var33_ult1
num_var2_ult1

imp_reemb_var33_ult1
num_aport_var33_ult1

imp_reemb_var33_ult1
num_var7_emit_ult1

imp_reemb_var33_ult1
num_meses_var13_medio_ult3

imp_reemb_var33_ult1
num_reemb_var13_hace3

imp_reemb_var33_ult1
num_reemb_var17_hace3

imp_reemb_var33_ult1
num_reemb_var17_ult1

imp_reemb_var33_ult1
num_reemb_var33_hace3

imp_reemb_var33_ult1
num_reemb_var33_ult1

imp_reemb_var33_ult1
num_trasp_var17_out_hace3

imp_reemb_var33_ult1
num_trasp_var17_out_ult1

imp_reemb_var33_ult1
num_trasp_var33_in_hace3

imp_reemb_var33_ult1
num_trasp_var33_out_hace3

imp_reemb_var33_ult1
saldo_var2_ult1

imp_reemb_var33_ult1
saldo_medio_var13_medio_hace2

imp_reemb_var33_ult1
saldo_medio_var13_medio_hace3

imp_reemb_var33_ult1
saldo_medio_var13_medio_ult1

imp_reemb

num_var7_emit_ult1
saldo_medio_var13_medio_hace2

num_var7_emit_ult1
saldo_medio_var13_medio_hace3

num_var7_emit_ult1
saldo_medio_var13_medio_ult1

num_var7_emit_ult1
saldo_medio_var13_medio_ult3

num_var7_emit_ult1
saldo_medio_var29_hace3

num_meses_var13_medio_ult3
num_reemb_var13_hace3

num_meses_var13_medio_ult3
num_reemb_var17_hace3

num_meses_var13_medio_ult3
num_reemb_var17_ult1

num_meses_var13_medio_ult3
num_reemb_var33_hace3

num_meses_var13_medio_ult3
num_reemb_var33_ult1

num_meses_var13_medio_ult3
num_trasp_var17_out_hace3

num_meses_var13_medio_ult3
num_trasp_var17_out_ult1

num_meses_var13_medio_ult3
num_trasp_var33_in_hace3

num_meses_var13_medio_ult3
num_trasp_var33_out_hace3

num_meses_var13_medio_ult3
saldo_var2_ult1

num_meses_var13_medio_ult3
saldo_medio_var13_medio_hace2

num_meses_var13_medio_ult3
saldo_medio_var13_medio_hace3

num_meses_var13_medio_ult3
saldo_medio_var13_medio_ult1

num_meses_var13_medio_ult3
saldo_medio_var13_medio_ult3

num_meses_var13_medio_

In [118]:
# let's check that indeed those features are duplicated
# I select a random pair from above

X_train[['ind_var2_0', 'num_var34_0']].head(10)

ind_var2_0  num_var34_0
10439           0            0
9236            0            0
818             0            0
11504           0            0
11722           0            0
5276            0            0
6863            0            0
13463           0            0
10228           0            0
11462           0            0

# Correlation

 Correlation Feature Selection evaluates subsets of features on the basis of the following hypothesis: "Good feature subsets contain features highly correlated with the target, yet uncorrelated to each other".

In [120]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn.model_selection import train_test_split

In [121]:
datasetc = pd.read_csv('D:/ShakthiSai -Keerthana Learnings/AI and Machine Learning/Feature Engineering/bnp-paribas-cardif-claims-management/BNP Paribas Cardif Claims Management-train.csv')
datasetc.shape

(114321, 133)

In [122]:
datasetc.head()

ID  target        v1        v2 v3        v4         v5        v6        v7  \
0   3       1  1.335739  8.727474  C  3.921026   7.915266  2.599278  3.176895   
1   4       1       NaN       NaN  C       NaN   9.191265       NaN       NaN   
2   5       1  0.943877  5.310079  C  4.410969   5.326159  3.979592  3.928571   
3   6       1  0.797415  8.304757  C  4.225930  11.627438  2.097700  1.987549   
4   8       1       NaN       NaN  C       NaN        NaN       NaN       NaN   

         v8  ...      v122      v123      v124  v125      v126      v127  \
0  0.012941  ...  8.000000  1.989780  0.035754    AU  1.804126  3.113719   
1  2.301630  ...       NaN       NaN  0.598896    AF       NaN       NaN   
2  0.019645  ...  9.333333  2.477596  0.013452    AE  1.773709  3.922193   
3  0.171947  ...  7.018256  1.812795  0.002267    CJ  1.415230  2.954381   
4       NaN  ...       NaN       NaN       NaN     Z       NaN       NaN   

       v128  v129      v130      v131  
0  2.024285     0  0.636365  2.857144  
1  1.957825     0       NaN       NaN  
2  1.120468     2  0.883118  1.176472  
3  1.990847     1  1.677108  1.034483  
4       NaN     0       NaN       NaN  

[5 rows x 133 columns]

In [123]:
# In practice, feature selection should be done after data pre-processing,
# so ideally, all the categorical variables are encoded into numbers,
# and then you can assess whether they are correlated with other features

# here for simplicity I will use only numerical variables
# select numerical columns:

numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
numerical_vars = list(datasetc.select_dtypes(include=numerics).columns)
datasetc = datasetc[numerical_vars]
datasetc.shape

(114321, 114)

In [124]:
X=datasetc.iloc[:,:-1]
y=datasetc.iloc[:,1]

In [ ]:
X

In [48]:
X_train, X_test, y_train, y_test = train_test_split(
    X,y,
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((80024, 113), (34297, 113))

In [49]:
corrmat = X_train.corr()

In [50]:
corrmat

ID    target        v1        v2        v4        v5        v6  \
ID      1.000000 -0.000094  0.001546  0.002185 -0.001768 -0.000218 -0.000017   
target -0.000094  1.000000 -0.011686  0.040219  0.061978  0.024694  0.032862   
v1      0.001546 -0.011686  1.000000 -0.217529 -0.141649 -0.065140 -0.017141   
v2      0.002185  0.040219 -0.217529  1.000000  0.539133  0.206560  0.021944   
v4     -0.001768  0.061978 -0.141649  0.539133  1.000000  0.270585  0.372836   
...          ...       ...       ...       ...       ...       ...       ...   
v126   -0.002281  0.003022  0.036250 -0.102610 -0.119823 -0.099690 -0.185749   
v127   -0.002551  0.010990 -0.005241  0.006894  0.297672 -0.219786  0.154535   
v128    0.001223  0.037691 -0.029843  0.268646  0.498675  0.592780  0.233378   
v129   -0.000784  0.143256 -0.023132  0.076418  0.108437  0.046271  0.045105   
v130    0.000517 -0.047777  0.259230 -0.577929 -0.699240 -0.257120 -0.298764   

              v7        v8        v9  ...      v120      v121      v122  \
ID     -0.000359  0.002066  0.006382  ...  0.006080 -0.003414  0.008253   
target  0.032442 -0.014400 -0.010910  ...  0.000070 -0.047611  0.004635   
v1      0.098819  0.152914 -0.024047  ...  0.048815  0.353783 -0.129889   
v2      0.349943 -0.361146 -0.011675  ... -0.109476 -0.574203  0.167348   
v4      0.473250 -0.164525 -0.208517  ... -0.038579 -0.615376  0.040830   
...          ...       ...       ...  ...       ...       ...       ...   
v126    0.181248  0.097838 -0.371714  ...  0.006057  0.140705 -0.399487   
v127    0.240705 -0.016341  0.132079  ...  0.316268  0.019691  0.104238   
v128    0.096187  0.014590 -0.282670  ... -0.185449 -0.380235 -0.058393   
v129    0.036604 -0.018991 -0.042973  ...  0.007637 -0.079281 -0.006749   
v130   -0.616537  0.313165  0.082964  ...  0.161499  0.806261 -0.188745   

            v123      v124      v126      v127      v128      v129      v130  
ID      0.001562  0.004479 -0.002281 -0.002551  0.001223 -0.000784  0.000517  
target -0.059299 -0.008996  0.003022  0.010990  0.037691  0.143256 -0.047777  
v1      0.300876  0.148090  0.036250 -0.005241 -0.029843 -0.023132  0.259230  
v2     -0.593202 -0.256963 -0.102610  0.006894  0.268646  0.076418 -0.577929  
v4     -0.710984 -0.048058 -0.119823  0.297672  0.498675  0.108437 -0.699240  
...          ...       ...       ...       ...       ...       ...       ...  
v126    0.130196  0.024044  1.000000 -0.082446 -0.142972 -0.013958  0.084175  
v127   -0.160184 -0.052421 -0.082446  1.000000 -0.154601  0.014115 -0.241227  
v128   -0.265147  0.257901 -0.142972 -0.154601  1.000000  0.061347 -0.296647  
v129   -0.098097 -0.010895 -0.013958  0.014115  0.061347  1.000000 -0.075994  
v130    0.732824  0.221640  0.084175 -0.241227 -0.296647 -0.075994  1.000000  

[113 rows x 113 columns]

In [51]:
corrmat=corrmat.abs().unstack()

In [52]:
corrmat

ID    ID        1.000000
      target    0.000094
      v1        0.001546
      v2        0.002185
      v4        0.001768
                  ...   
v130  v126      0.084175
      v127      0.241227
      v128      0.296647
      v129      0.075994
      v130      1.000000
Length: 12769, dtype: float64

In [53]:
corrmat = corrmat.sort_values(ascending=False)

In [54]:
corrmat

v130  v130    1.000000
v82   v82     1.000000
v94   v94     1.000000
v93   v93     1.000000
v92   v92     1.000000
                ...   
v126  v80     0.000028
v6    ID      0.000017
ID    v6      0.000017
      v93     0.000004
v93   ID      0.000004
Length: 12769, dtype: float64

In [55]:
corrmat = corrmat[corrmat >= 0.8]
corrmat = corrmat[corrmat < 1]

In [57]:
corrmat

v100  v58     0.997626
v58   v100    0.997626
v115  v69     0.993510
v69   v115    0.993510
v76   v17     0.992448
                ...   
v17   v101    0.801587
v78   v16     0.801199
v16   v78     0.801199
v124  v46     0.800613
v46   v124    0.800613
Length: 248, dtype: float64

In [58]:
corrmat = pd.DataFrame(corrmat).reset_index()
corrmat.columns = ['feature1', 'feature2', 'corr']
corrmat.head()

feature1 feature2      corr
0     v100      v58  0.997626
1      v58     v100  0.997626
2     v115      v69  0.993510
3      v69     v115  0.993510
4      v76      v17  0.992448